File: haskell/problem3.hs
isqrt :: Integer -> Integer
isqrt x = floor $ sqrt $ fromIntegral x
is_prime :: Integer -> Bool
is_prime k = null [ x | x <- [2..isqrt k], k `mod`x  == 0]
source = 600851475143
sq = isqrt source
maximum [ x | x <- [3..sq], is_prime x, source `mod` x == 0]

File: go/problem3.go
package main

import (
	"fmt"
	"math"
)

func is_prime(n int) bool {
	for j := 2; j < n; j++ {
		if n%j == 0 {
			break
		} else if j == n-1 {
			return true
		}
	}
	return false
}

func main() {
	target := 600851475143
	max := int(math.Floor(math.Sqrt(600851475143)))
	var candidate int
	for i := 1; i < max; i++ {
		if target%i == 0 {
			if is_prime(i) {
				candidate = i
			}
		}
	}
	fmt.Println("Winner is:", candidate)
}
